### PR TITLE
Update RAP implementation

### PR DESCRIPTION
**Pull request by enver1323**
This PR introduces updated RAP implementation. To make the RAP fully adaptive, attention mechanism and indexing are needed. Other that that, method should work without any issues.

### CODE REVIEW

**Review by ghkim9213 — COMMENTED** [review]
Thanks for your hard work, fx master. I've got many hints for the usage of fx from your codes.

Actually, I tested this version of rap and I found some issues as followings
- Heatmaps are too much different from the previous version while the codes look fine

prev rap: resnet18-tench
prev_rap

curr rap: resnet18-tench
curr_rap

- Still not work with inception_v3
emsg_inceptoin

Do you have same issues?

p.s. I'm trying to build rap based on zennit on the branch feature/rap-by-zennit. Currently, it looks working good for linear layers including conv, while not for batchnorm (or add). I'm trying to fix and I'll let you know if I have any progress.

**Comment by enver1323** [comment]
Thank you so much, @ghkim9213 . I will try to fix the heatmaps to be the same. I have also similar results. I looked into inception v3. The issue is related to `index` method in preprocessing step of the model, e.g. `a[:, 0]`

**Comment by enver1323** [comment]
I tested implementation of original RAP package, and it produces similar random results on resnet18 

image

I am not sure about the reason behind such behavior. Authors test correctness of explanations by comparing sums of prediction logits and relevances. I conducted the same test on this implementation version. The results are as follows:
```
Pred logit : 31.355219
Relevance Sum : 31.355202
Pred logit : 14.111381
Relevance Sum : 14.111369
Pred logit : 16.852928
Relevance Sum : 16.85293
Pred logit : 20.30917
Relevance Sum : 20.309172
Pred logit : 12.61315
Relevance Sum : 12.613142
```
One more issue, I found during the test is that `Pred logit` and `Relevance Sum` values do not match when tensor is batched. However, as it can be seen by the results above, values match, when processed one-by-one.

I will investigate further, just wanted to share the results of my findings.

**Comment by enver1323** [comment]
The last commit fixed an issue with multiple items in a batch

**Comment by ghkim9213** [comment]
> I tested implementation of original RAP package, and it produces similar random results on resnet18
> 
> image
> 
> I am not sure about the reason behind such behavior. Authors test correctness of explanations by comparing sums of prediction logits and relevances. I conducted the same test on this implementation version. The results are as follows:
> 
> ```
> Pred logit : 31.355219
> Relevance Sum : 31.355202
> Pred logit : 14.111381
> Relevance Sum : 14.111369
> Pred logit : 16.852928
> Relevance Sum : 16.85293
> Pred logit : 20.30917
> Relevance Sum : 20.309172
> Pred logit : 12.61315
> Relevance Sum : 12.613142
> ```
> 
> One more issue, I found during the test is that `Pred logit` and `Relevance Sum` values do not match when tensor is batched. However, as it can be seen by the results above, values match, when processed one-by-one.
> 
> I will investigate further, just wanted to share the results of my findings.

Same issues of the original package on my env. Just for informing.
- tench-vgg
rap_author_tench_vgg

- tench-resnet18
rap_author_tench_resnet18

- tench-resnet50
rap_author_tench_resnet50

I guess the author failed to process relevance for the residual connection for the torch model. In the paper, they said they use keras model instead of torch one.

https://github.com/wjNam/Relative_Attributing_Propagation/issues/6

**Comment by enver1323** [comment]
If you do not mind, I will merge existing work to the main, so that we had functioning RAP for now. We can update it, when all issues a resolved.

**Comment by enver1323** [comment]
For some reason applying relu before adding block improves explanation quality substantially
newplot

**Review by ghkim9213 — APPROVED** [review]
